bpo-38971: open file in codecs.open() closes if exception raised

caporta opened the pull request:
https://bugs.python.org/issue38971

Open issue in the BPO indicated a desire to make the implementation of
codecs.open() at parity with io.open(), which implements a try/except to
assure file stream gets closed before an exception is raised.


<!-- issue-number: [bpo-38971](https://bugs.python.org/issue38971) -->
https://bugs.python.org/issue38971
<!-- /issue-number -->

caporta committed:
bpo-38971: open file in codecs.open() closes if exception raised

Open issue in the BPO indicated a desire to make the implementation of
codecs.open() at parity with io.open(), which implements a try/except to
assure file stream gets closed before an exception is raised.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

serhiy-storchaka commented on Lib/test/test_codecs.py:
```diff
@@ -1154,6 +1154,19 @@ def test_stream_bare(self):
             got = ostream.getvalue()
             self.assertEqual(got, unistring)
 
+class FileClosesOnRaiseTest(unittest.TestCase):
```
I think there is no need to add a new class for this test. You can add it in `CodecsModuleTest`.

serhiy-storchaka commented on Lib/test/test_codecs.py:
```diff
@@ -1154,6 +1154,19 @@ def test_stream_bare(self):
+        mock_open = mock.mock_open()
+        mock_lookup = mock.Mock(side_effect=Exception)
+
+        with mock.patch('codecs.lookup', mock_lookup):
```
Instead of patching `codecs.lookup`, just pass an invalid encoding.

serhiy-storchaka commented on Lib/test/test_codecs.py:
```diff
@@ -1154,6 +1154,19 @@ def test_stream_bare(self):
+
+        with mock.patch('codecs.lookup', mock_lookup):
+            with mock.patch('builtins.open', mock_open) as file:
+                with self.assertRaises(Exception):
```
`Exception` is too wide. Check that more specific `LookupError` is raised.

serhiy-storchaka reviewed (commented):
The fix LGTM, but I have some comments about the test.

caporta committed:
Improve test composition

- Move spec out of seperate class and into CodecsModuleTest
- Pass invalid encoding to force exception rather than mock
- Specify that a LookupError is raised specifically

serhiy-storchaka commented on Lib/test/test_codecs.py:
```diff
@@ -1713,6 +1714,14 @@ def test_undefined(self):
+        mock_open = mock.mock_open()
+        with mock.patch('builtins.open', mock_open) as file:
+            with self.assertRaises(LookupError):
+                codecs.open(support.TESTFN, mode = 'wt', encoding='invalid-format')
```
Too long line and spaces around "=".

Actually using keywords for passing argument is not needed.

```suggestion
                codecs.open(support.TESTFN, 'wt', 'invalid-encoding')
```

caporta replied:
Thanks! Committed the suggested change above.

serhiy-storchaka reviewed (commented):
Thank you. Yet a nitpick to tests.

caporta committed:
Update Lib/test/test_codecs.py

Prefer positional args to keywords in spec

Co-Authored-By: Serhiy Storchaka <storchaka@gmail.com>
